## Proyecto Pandas 
##### Jacobo Maté

In [1]:
import numpy as np # Importamos paquetes
import pandas as pd

In [2]:
path = '../data/attacks.csv' # Importamos base de datos
data = pd.read_csv(path, encoding = "ISO-8859-1")

In [3]:
pd.set_option('display.max_columns', None)  # ver todas las columnas

pd.set_option('display.max_rows', None)     # ver todas las filas

In [4]:
data.head() # Mostramos la cabeza de los datos

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   Age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                    Habush, Oahu      Surfing      John Denges    M   48   
3              Arrawarra Headland      Surfing             male    M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos     M  NaN   

                                              Injury Fatal (Y/N)  \
0  No injury to occupant, outrigger canoe and pad...           N   
1                         Minor injury to left thigh           N   
2       Injury to left lower leg from surfboard skeg           N   
3                          Minor injury to lower leg           N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT           N   

            Time         Species           Investigator or Source  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2  original order Unnamed: 22 Unnamed: 23  
0    2018.06.25          6303.0         NaN         NaN  
1    2018.06.18          6302.0         NaN         NaN  
2    2018.06.09          6301.0         NaN         NaN  
3    2018.06.08          6300.0         NaN         NaN  
4    2018.06.04          6299.0         NaN         NaN

#### ¿Qué podemos sacar de este head?
 * Tenemos varias columnas con título con mayúsculas y espacios, voy a cambiarlas todas a lower y hacer replace de espacios. Esto es útil de cara a más tarde poder usar punto en vez de corchetes y comillado al querer referirme a una columna

In [5]:
data.columns = [c.lower().replace(' ', '') for c in data.columns] # Paso a minúsculas y quito los espacios para poder usar puntos
data.columns

Index(['casenumber', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex', 'age', 'injury', 'fatal(y/n)', 'time',
       'species', 'investigatororsource', 'pdf', 'hrefformula', 'href',
       'casenumber.1', 'casenumber.2', 'originalorder', 'unnamed:22',
       'unnamed:23'],
      dtype='object')

#### Limpieza | Ordenado datos `nans` y `duplicates`

In [6]:
data.info() # Procedemos a imprimir el tipo de datos con el que vamos a trabajar

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   casenumber            8702 non-null   object 
 1   date                  6302 non-null   object 
 2   year                  6300 non-null   float64
 3   type                  6298 non-null   object 
 4   country               6252 non-null   object 
 5   area                  5847 non-null   object 
 6   location              5762 non-null   object 
 7   activity              5758 non-null   object 
 8   name                  6092 non-null   object 
 9   sex                   5737 non-null   object 
 10  age                   3471 non-null   object 
 11  injury                6274 non-null   object 
 12  fatal(y/n)            5763 non-null   object 
 13  time                  2948 non-null   object 
 14  species               3464 non-null   object 
 15  investigatororsourc

In [7]:
data.shape

(25723, 24)

Tenemos un total de 25723 filas y 24 columnas

#### ¿Qué podemos sacar ya del tipo de datos con los que vamos a trabajar?
* Tenemos un total de 22 columnas de tipo objeto - Serán cadenas de caracteres
* Tenemos dos columnas de tipo float - columnas `year` y `originalorder`
* Es raro que un año sea de tipo float -> No va a haber años float tipo 2023.5


In [8]:
# Queremos cambiar el tipo de dato de year a int.
# Para ello primero comprobamos si hay algun nan con any() - Sacamos la suma de nulos que hay en cada columna
data.isnull().sum()

casenumber              17021
date                    19421
year                    19423
type                    19425
country                 19471
area                    19876
location                19961
activity                19965
name                    19631
sex                     19986
age                     22252
injury                  19449
fatal(y/n)              19960
time                    22775
species                 22259
investigatororsource    19438
pdf                     19421
hrefformula             19422
href                    19421
casenumber.1            19421
casenumber.2            19421
originalorder           19414
unnamed:22              25722
unnamed:23              25721
dtype: int64

* En este output podemos ver la gran cantidad de valores nulos que hay. A simple vista se ve que mínimo hay 17000 nulos por columna. Coincide que en varias columnas hay exactamente 19421 nulos. 
* Partiendo de la hipótesis de que esos 19421 datos nulos van a concidir en fila con los otros, voy a decidir borrar dichas filas. 
* Para ello voy a tomar como número de nulos mas restrictivo el más pequeño 17021 - columna `casenumber` - sacaré los índices de dichas filas y borraré del database sus filas.

In [9]:
indices_malos = data[data.casenumber.isna() == True].index # Aquí usamos un filtro para almacenar en la variable indices_malos los indices de las filas de la columna casenumber que sean nulos
indices_malos

Index([ 5488,  8702,  8703,  8704,  8705,  8706,  8707,  8708,  8709,  8710,
       ...
       25712, 25713, 25714, 25715, 25716, 25717, 25718, 25719, 25720, 25721],
      dtype='int64', length=17021)

* Voy a revisar individualmente el caso del índice 5488 porque el resto parece que son seguidas e intuyo que el resto de valores tb sean nulos para ese rango 8702-25721

In [10]:
data[data.index == 5488]

casenumber                  date    year      type country     area  \
5488        NaN  Reported 06-Sep-1905  1905.0  Provoked     USA  Florida   

                          location activity         name sex  age  \
5488  Fort Pierce, St Lucie County  Fishing  David Curry   M  NaN   

                                                 injury fatal(y/n) time  \
5488  Lacerations to leg from hooked shark PROVOKED ...          N  NaN   

     species            investigatororsource                     pdf  \
5488     NaN  Muncie Evening Press, 8/6/1913  1905.09.06.R-Curry.pdf   

                                            hrefformula  \
5488  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href  casenumber.1  \
5488  http://sharkattackfile.net/spreadsheets/pdf_di...  1905.09.06.R   

      casenumber.2  originalorder unnamed:22 unnamed:23  
5488  1905.09.06.R          815.0        NaN        NaN

In [11]:
# Como si que hay datos para este ataque me lo voy a quedar. 
# Sí que voy a borrar los datos de las filas correspondientes al rango 8702-25721.

data.drop([i for i in range(8702,25721)],axis=0).shape

(8704, 24)

In [12]:
# Antes de aplicar el cambio veo que esté haciéndolo bien imprimiendo el shape. Hemos bajado las filas a 8704 así que bien -> Procedo con el inplace=True
data.drop([i for i in range(8702,25722)],axis=0,inplace=True)
data.shape

(8703, 24)

In [13]:
data.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [14]:
data.isnull().sum()

casenumber                 1
date                    2401
year                    2403
type                    2405
country                 2451
area                    2856
location                2941
activity                2945
name                    2611
sex                     2966
age                     5232
injury                  2429
fatal(y/n)              2940
time                    5755
species                 5239
investigatororsource    2418
pdf                     2401
hrefformula             2402
href                    2401
casenumber.1            2401
casenumber.2            2401
originalorder           2394
unnamed:22              8702
unnamed:23              8701
dtype: int64

Volviendo al punto donde nos había extrañado que el tipo de `year` fuera float; si queremos cambiar el tipo de esta columna a int tendremos que cambiar los nans a un valor -> sino al aplicar el astype() nos saltará un error.

Para ello voy a proceder como antes: saco los indices de las filas donde se encuentren valores nulos en la columna `year` -> Exploro tabla para esos índices y si nada me indica lo contrario cambio ese nan a algo con lo que pueda trabajar

In [15]:
yearsnulos = data[data.year.isna() == True].index
yearsnulos

Index([  187,  6079,  6302,  6303,  6304,  6305,  6306,  6307,  6308,  6309,
       ...
        8693,  8694,  8695,  8696,  8697,  8698,  8699,  8700,  8701, 25722],
      dtype='int64', length=2403)

In [16]:
data.loc[yearsnulos].head(10)

casenumber                  date  year        type    country  \
187   2017.01.08.R  Reported 08-Jan-2017   NaN     Invalid  AUSTRALIA   
6079  1836.08.19.R  Reported 19-Aug-1836   NaN  Unprovoked    ENGLAND   
6302             0                   NaN   NaN         NaN        NaN   
6303             0                   NaN   NaN         NaN        NaN   
6304             0                   NaN   NaN         NaN        NaN   
6305             0                   NaN   NaN         NaN        NaN   
6306             0                   NaN   NaN         NaN        NaN   
6307             0                   NaN   NaN         NaN        NaN   
6308             0                   NaN   NaN         NaN        NaN   
6309             0                   NaN   NaN         NaN        NaN   

            area    location      activity          name  sex  age  \
187   Queensland         NaN  Spearfishing  Kerry Daniel    M   35   
6079  Cumberland  Whitehaven      Swimming         a boy    M  NaN   
6302         NaN         NaN           NaN           NaN  NaN  NaN   
6303         NaN         NaN           NaN           NaN  NaN  NaN   
6304         NaN         NaN           NaN           NaN  NaN  NaN   
6305         NaN         NaN           NaN           NaN  NaN  NaN   
6306         NaN         NaN           NaN           NaN  NaN  NaN   
6307         NaN         NaN           NaN           NaN  NaN  NaN   
6308         NaN         NaN           NaN           NaN  NaN  NaN   
6309         NaN         NaN           NaN           NaN  NaN  NaN   

                                      injury fatal(y/n) time     species  \
187   No attack, shark made a threat display        NaN  NaN  Bull shark   
6079                                   FATAL          Y  NaN         NaN   
6302                                     NaN        NaN  NaN         NaN   
6303                                     NaN        NaN  NaN         NaN   
6304                                     NaN        NaN  NaN         NaN   
6305                                     NaN        NaN  NaN         NaN   
6306                                     NaN        NaN  NaN         NaN   
6307                                     NaN        NaN  NaN         NaN   
6308                                     NaN        NaN  NaN         NaN   
6309                                     NaN        NaN  NaN         NaN   

        investigatororsource                           pdf  \
187   Liquid Vision 1/8/2017  2017.01.08.R-KerryDaniel.pdf   
6079          C. Moore, GSAF   1835.08.19.R-Whitehaven.pdf   
6302                     NaN                           NaN   
6303                     NaN                           NaN   
6304                     NaN                           NaN   
6305                     NaN                           NaN   
6306                     NaN                           NaN   
6307                     NaN                           NaN   
6308                     NaN                           NaN   
6309                     NaN                           NaN   

                                            hrefformula  \
187   http://sharkattackfile.net/spreadsheets/pdf_di...   
6079  http://sharkattackfile.net/spreadsheets/pdf_di...   
6302                                                NaN   
6303                                                NaN   
6304                                                NaN   
6305                                                NaN   
6306                                                NaN   
6307                                                NaN   
6308                                                NaN   
6309                                                NaN   

                                                   href  casenumber.1  \
187   http://sharkattackfile.net/spreadsheets/pdf_di...  2017.01.08.R   
6079  http://sharkattackfile.net/spreadsheets/pdf_di...  1836.08.19.R   
6302                                        

In [17]:
data.loc[yearsnulos].tail(10)

casenumber date  year type country area location activity name  sex  \
8693           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8694           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8695           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8696           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8697           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8698           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8699           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8700           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
8701           0  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   
25722         xx  NaN   NaN  NaN     NaN  NaN      NaN      NaN  NaN  NaN   

       age injury fatal(y/n) time species investigatororsource  pdf  \
8693   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8694   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8695   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8696   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8697   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8698   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8699   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8700   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
8701   NaN    NaN        NaN  NaN     NaN                  NaN  NaN   
25722  NaN    NaN        NaN  NaN     NaN                  NaN  NaN   

      hrefformula href casenumber.1 casenumber.2  originalorder unnamed:22  \
8693          NaN  NaN          NaN          NaN            NaN        NaN   
8694          NaN  NaN          NaN          NaN            NaN        NaN   
8695          NaN  NaN          NaN          NaN            NaN        NaN   
8696          NaN  NaN          NaN          NaN            NaN        NaN   
8697          NaN  NaN          NaN          NaN            NaN        NaN   
8698          NaN  NaN          NaN          NaN            NaN        NaN   
8699          NaN  NaN          NaN          NaN            NaN        NaN   
8700          NaN  NaN          NaN          NaN            NaN        NaN   
8701          NaN  NaN          NaN          NaN            NaN        NaN   
25722         NaN  NaN          NaN          NaN            NaN        NaN   

      unnamed:23  
8693         NaN  
8694         NaN  
8695         NaN  
8696         NaN  
8697         NaN  
8698         NaN  
8699         NaN  
8700         NaN  
8701         NaN  
25722        NaN

* Vista la salida, intuyo que desde la fila 6303 hasta 8701 tenemos valores nulos en todas las columnas salvo en `originalorder`.
* Como de 6302 a 8701 hay 2400 filas y sabemos que en total hay 2403 nulos, podemos confirmar que ese rango lo podemos eliminar todo también.
* Si todos los valores son nulos menos una columna tampoco me sirven de mucho -> Puedo hacer un último checkeo para comprobar que mi hipótesis sea verdadera y que no me esté fumando datos correctos por ahí enmedio.

In [18]:
data.drop([i for i in range(6303,8701)],axis=0).shape

(6305, 24)

In [19]:
data.drop([i for i in range(6303,8701)],axis=0, inplace=True)
data.shape

(6305, 24)

In [20]:
data.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [21]:
# Hacemos también un tail por si queda alguna fila nan al final del databse:
data.tail(10)

casenumber         date  year        type             country  \
6295     ND.0007  Before 1906   0.0  Unprovoked           AUSTRALIA   
6296     ND.0006  Before 1906   0.0  Unprovoked           AUSTRALIA   
6297     ND.0005  Before 1903   0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903   0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905   0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889   0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853   0.0  Unprovoked  CEYLON (SRI LANKA)   
6302           0          NaN   NaN         NaN                 NaN   
8701           0          NaN   NaN         NaN                 NaN   
25722         xx          NaN   NaN         NaN                 NaN   

                    area                             location      activity  \
6295                 NaN                                  NaN       Fishing   
6296     New South Wales                                           Swimming   
6297   Western Australia                          Roebuck Bay        Diving   
6298   Western Australia                                  NaN  Pearl diving   
6299      North Carolina                       Ocracoke Inlet      Swimming   
6300                 NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301    Eastern Province  Below the English fort, Trincomalee      Swimming   
6302                 NaN                                  NaN           NaN   
8701                 NaN                                  NaN           NaN   
25722                NaN                                  NaN           NaN   

                        name  sex  age  \
6295               fisherman    M  NaN   
6296                Arab boy    M  NaN   
6297                    male    M  NaN   
6298                   Ahmun    M  NaN   
6299   Coast Guard personnel    M  NaN   
6300         Jules Patterson    M  NaN   
6301                    male    M   15   
6302                     NaN  NaN  NaN   
8701                     NaN  NaN  NaN   
25722                    NaN  NaN  NaN   

                                                  injury fatal(y/n) time  \
6295                                               FATAL          Y  NaN   
6296                                               FATAL          Y  NaN   
6297                                               FATAL          Y  NaN   
6298                                               FATAL          Y  NaN   
6299                                               FATAL          Y  NaN   
6300                                               FATAL          Y  NaN   
6301   FATAL. "Shark bit him in half, carrying away t...          Y  NaN   
6302                                                 NaN        NaN  NaN   
8701                                                 NaN        NaN  NaN   
25722                                                NaN        NaN  NaN   

                                                 species  \
6295                                       Blue pointers   
6296   Said to involve a grey nurse shark that leapt ...   
6297                                                 NaN   
6298                                                 NaN   
6299                                                 NaN   
6300                                                 NaN   
6301                                                 NaN   
6302                                                 NaN   
8701                                                 NaN   
25722                                                NaN   

                                    investigatororsource  \
6295   NY Sun, 9/9/1906, referring to account by Loui...   
6296   L. Becke in New York Sun, 9/9/1906; L. Schultz...   
6297                    H. Taunton; N. Bartlett,  p. 234   
6298               H. Taunton; N. Bartlett,  pp. 233-234   
6299                F. Schwartz, p.23; C. Creswell, GSAF   
6300                             

In [22]:
# Borramos también esas tres filas de nans enteras:
data.drop([6302,8701,25722],axis=0, inplace=True)

In [23]:
data.shape

(6302, 24)

In [24]:
data.isnull().sum()

casenumber                 1
date                       0
year                       2
type                       4
country                   50
area                     455
location                 540
activity                 544
name                     210
sex                      565
age                     2831
injury                    28
fatal(y/n)               539
time                    3354
species                 2838
investigatororsource      17
pdf                        0
hrefformula                1
href                       0
casenumber.1               0
casenumber.2               0
originalorder              0
unnamed:22              6301
unnamed:23              6300
dtype: int64

* Después de borrar los nans del paso anterior vuelvo a mostrar la suma de estos.
* A partir de aquí no vamos a quitar más filas xq entendemos que los valores nulos serán por falta de información - no siempre se sabrá especie de tiburón, edad, actividad que estaba practicando la víctima...
* Ahora todo lo que sea nan lo vamos a pasar a unknown utilizando el fillna()

In [25]:
data.fillna('unknown').tail()

casenumber         date year        type             country  \
6297    ND.0005  Before 1903  0.0  Unprovoked           AUSTRALIA   
6298    ND.0004  Before 1903  0.0  Unprovoked           AUSTRALIA   
6299    ND.0003    1900-1905  0.0  Unprovoked                 USA   
6300    ND.0002    1883-1889  0.0  Unprovoked              PANAMA   
6301    ND.0001    1845-1853  0.0  Unprovoked  CEYLON (SRI LANKA)   

                   area                             location      activity  \
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                              unknown  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300            unknown                 Panama Bay 8ºN, 79ºW       unknown   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       name sex      age  \
6297                   male   M  unknown   
6298                  Ahmun   M  unknown   
6299  Coast Guard personnel   M  unknown   
6300        Jules Patterson   M  unknown   
6301                   male   M       15   

                                                 injury fatal(y/n)     time  \
6297                                              FATAL          Y  unknown   
6298                                              FATAL          Y  unknown   
6299                                              FATAL          Y  unknown   
6300                                              FATAL          Y  unknown   
6301  FATAL. "Shark bit him in half, carrying away t...          Y  unknown   

      species                   investigatororsource  \
6297  unknown       H. Taunton; N. Bartlett,  p. 234   
6298  unknown  H. Taunton; N. Bartlett,  pp. 233-234   
6299  unknown   F. Schwartz, p.23; C. Creswell, GSAF   
6300  unknown                    The Sun, 10/20/1938   
6301  unknown                             S.W. Baker   

                                 pdf  \
6297          ND-0005-RoebuckBay.pdf   
6298               ND-0004-Ahmun.pdf   
6299  ND-0003-Ocracoke_1900-1905.pdf   
6300      ND-0002-JulesPatterson.pdf   
6301              ND-0001-Ceylon.pdf   

                                            hrefformula  \
6297  http://sharkattackfile.net/spreadsheets/pdf_di...   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...   
6299  http://sharkattackfile.net/spreadsheets/pdf_di...   
6300  http://sharkattackfile.net/spreadsheets/pdf_di...   
6301  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href casenumber.1  \
6297  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0005   
6298  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0004   
6299  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0003   
6300  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0002   
6301  http://sharkattackfile.net/spreadsheets/pdf_di...      ND.0001   

     casenumber.2  originalorder unnamed:22 unnamed:23  
6297      ND.0005            6.0    unknown    unknown  
6298      ND.0004            5.0    unknown    unknown  
6299      ND.0003            4.0    unknown    unknown  
6300      ND.0002            3.0    unknown    unknown  
6301      ND.0001            2.0    unknown    unknown

* Nos falta hacer comprobar si hay duplicados - procedemos a ello

In [26]:
data.drop_duplicates(inplace=True)

In [27]:
data.shape

(6302, 24)

Tenemos la misma dimensión de filas por lo que asumimos que no había duplicados.

In [28]:
data.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

In [29]:
# Hago una copia de la base de datos para tener un chackpoint al que poder volver:
dc = data.copy() # dc por data clean

In [30]:
dc.shape

(6302, 24)

In [31]:
dc.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

#### Limpieza | Ordenado columnas `type` y `activity`

* Voy a continuar el estudio por obteniendo los valores únicos de la columna type. Ya que no se incluye una descripción de estos datos en la fuente de la base de datos vamos a ver qué tipos de detalles se incluyen en esta columna y si estos difieren mucho de un suceso a otro.

In [32]:
dc.type.unique()

array(['Boating', 'Unprovoked', 'Invalid', 'Provoked', 'Questionable',
       'Sea Disaster', nan, 'Boat', 'Boatomg'], dtype=object)

In [33]:
dc.type.value_counts()

type
Unprovoked      4595
Provoked         574
Invalid          547
Sea Disaster     239
Boating          203
Boat             137
Questionable       2
Boatomg            1
Name: count, dtype: int64

In [34]:
# Vamos a cambiar la de tipo 'Boatomg' por tipo Boating ya que creo se trata de una errata
ind_errata = dc[dc.type == 'Boatomg'].index
ind_errata

Index([1152], dtype='int64')

In [35]:
# Cambiamos el dato de type del indice 1152 por Boating
dc.type[dc.index==1152] = 'Boating'


C:\Users\jacom\AppData\Local\Temp\ipykernel_7860\1426339824.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dc.type[dc.index==1152] = 'Boating'


In [36]:
dc[dc.index == 1152]

casenumber         date    year     type      country          area  \
1152  2009.03.01.b  01-Mar-2009  2009.0  Boating  NEW ZEALAND  North Island   

      location activity                                             name sex  \
1152  Taranaki  Fishing  boat, occupants: Boyd Rutherford & Hamish Roper   M   

      age                                       injury fatal(y/n) time  \
1152  NaN  No injury to occupants, shark bit propeller          N  NaN   

     species           investigatororsource                      pdf  \
1152     NaN  Taranaki Daily News, 3/3/2009  2009.03.01-Taranaki.pdf   

                                            hrefformula  \
1152  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href  casenumber.1  \
1152  http://sharkattackfile.net/spreadsheets/pdf_di...  2009.03.01.b   

      casenumber.2  originalorder unnamed:22 unnamed:23  
1152  2009.03.01.b         5151.0        NaN        NaN

In [37]:
dc.type.value_counts()

type
Unprovoked      4595
Provoked         574
Invalid          547
Sea Disaster     239
Boating          204
Boat             137
Questionable       2
Name: count, dtype: int64

In [38]:
# También vamos a meter en una sola categoría los datos de boating y boat -> Boating
ind_boat = dc[dc.type == 'Boat'].index
ind_boat

Index([ 538,  565,  706,  818,  864,  956,  971, 1164, 1168, 1174,
       ...
       4188, 4214, 4242, 6070, 6087, 6127, 6136, 6145, 6212, 6274],
      dtype='int64', length=137)

In [39]:
# Cambiamos el dato de type del indice 1152 por Boating
dc.type.loc[ind_boat] = 'Boating'

C:\Users\jacom\AppData\Local\Temp\ipykernel_7860\4105221077.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dc.type.loc[ind_boat] = 'Boating'


In [40]:
dc.type.value_counts()

type
Unprovoked      4595
Provoked         574
Invalid          547
Boating          341
Sea Disaster     239
Questionable       2
Name: count, dtype: int64

In [41]:
dc.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location     activity             name sex  age  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely    F   11   
2                    Habush, Oahu      Surfing      John Denges   M   48   
3              Arrawarra Headland      Surfing             male   M  NaN   
4                        La Ticla  Free diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2          07h45              NaN  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN

+ Dado que es la víctima la que decide si el ataque ha sido provocado o no, en mi opinión es un valor que no debería tenerse en cuenta ya que me parece totalmente sesgado.
* Seguimos con la actividad que practicaba la víctima del ataque

In [42]:
actividades = pd.DataFrame(dc.activity.value_counts())

In [43]:
actividades.shape

(1532, 1)

In [44]:
actividades.head(50)

count
activity                             
Surfing                           971
Swimming                          869
Fishing                           431
Spearfishing                      333
Bathing                           162
Wading                            149
Diving                            127
Standing                           99
Snorkeling                         89
Scuba diving                       76
Body boarding                      61
Body surfing                       49
Swimming                           47
Kayaking                           33
Treading water                     32
Pearl diving                       32
Fell overboard                     32
Free diving                        29
Boogie boarding                    29
Windsurfing                        19
Walking                            17
Boogie Boarding                    16
Shark fishing                      15
Floating                           14
Fishing                            13
Canoeing                           13
Surf skiing                        12
Surf fishing                       12
Surf-skiing                        12
Rowing                             12
Fishing for sharks                 11
Kayak Fishing                      11
Scuba Diving                       10
Freediving                         10
Sponge diving                      10
Fell into the water                 9
Sailing                             9
Sitting on surfboard                9
Diving for trochus                  9
Paddle boarding                     8
Sea disaster                        8
Diving for abalone                  7
Free diving for abalone             7
Skindiving                          7
Playing                             7
Surf skiing                         7
Surfing (sitting on his board)      7
Boating                             7
Floating on his back                7
Spearfishing                        7

Podemos ver que hay 1532 actividades diferentes - sin embargo muchas de ellas se podrian abarcar bajo campos parecidos

Algunos grandes campos pueden ser los siguientes:
* Fishing
* Swimming
* Surfing (any kind of board sport)
* Diving

Voy a llevar a cabo un estudio y actualizado de campos utilizando regex. 

In [45]:
import re # Importo paquete regex

# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_swim(cadena):

    patron = r'swim' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena, re.IGNORECASE)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Swimming'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [46]:
res = busca_swim('SWImidning in da pool') # Compruebo que funcione
print(res)

Swimming


In [47]:
res = busca_swim('cadena sin palabra')
print(res)

cadena sin palabra


In [48]:
# Ahora voy a hacer esta función exactamente igual para los otros casos de diving, fishing y surfing

In [49]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_dive(cadena):

    cadena = cadena.lower()

    patron = r'div' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Diving'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [50]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_fish(cadena):

    cadena = cadena.lower()

    patron = r'fish' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Fishing'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [51]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_surf(cadena):

    cadena = cadena.lower()

    patron = r'surf' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Surfing'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [52]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_over(cadena):

    cadena = cadena.lower()

    patron = r'over' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Overboard or capsized'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

In [53]:
# Defino una función que  busque los caracteres clave en la cadena de entrada
def busca_board(cadena): # Las actividades que sean de tipo board las voy a meter en surfing también

    cadena = cadena.lower()

    patron = r'board' # Defino un patrón de busqueda

    resultado = re.search(patron, cadena)

    if resultado: # Si el resultado no está vacío quiere decir que lo ha encontrado
        return 'Surfing'
    else:
        return cadena # En caso de que no la encuentre se devuelve cadena original

Procedo a aplicar las 4 fórmulas que he definido a la columna `activity` con el formato apply

In [54]:
dc.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6302 entries, 0 to 6301
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   casenumber            6301 non-null   object 
 1   date                  6302 non-null   object 
 2   year                  6300 non-null   float64
 3   type                  6298 non-null   object 
 4   country               6252 non-null   object 
 5   area                  5847 non-null   object 
 6   location              5762 non-null   object 
 7   activity              5758 non-null   object 
 8   name                  6092 non-null   object 
 9   sex                   5737 non-null   object 
 10  age                   3471 non-null   object 
 11  injury                6274 non-null   object 
 12  fatal(y/n)            5763 non-null   object 
 13  time                  2948 non-null   object 
 14  species               3464 non-null   object 
 15  investigatororsource  6285

In [55]:
dc.activity = dc.activity.astype(str)

In [56]:
# Procedo a aplicar las 4 fórmulas a la columna activity del dataframe
dc.activity = dc.activity.apply(busca_swim)
dc.activity = dc.activity.apply(busca_dive)
dc.activity = dc.activity.apply(busca_surf)
dc.activity = dc.activity.apply(busca_fish)
dc.activity = dc.activity.apply(busca_over) # Llamo primero a esta función para quitarme los overboard antes de los deportes de tabla
dc.activity = dc.activity.apply(busca_board)

In [57]:
actividades_limpia = pd.DataFrame(dc.activity.value_counts())
actividades_limpia.head(150)

count
activity                                                 
surfing                                              1219
fishing                                              1125
swimming                                             1123
diving                                                600
nan                                                   544
Surfing                                               279
bathing                                               162
wading                                                149
standing                                               99
snorkeling                                             89
kayaking                                               33
treading water                                         32
walking                                                17
floating                                               14
canoeing                                               14
sea disaster                                           14
rowing                                                 12
sailing                                                 9
fell into the water                                     9
floating on his back                                    7
playing                                                 7
boating                                                 7
shipwreck                                               6
murder                                                  6
dangling feet in the water                              5
paddleskiing                                            5
boat                                                    5
boat capsized                                           4
jumping                                                 4
splashing                                               4
seine netting                                           4
jumped into the water                                   4
tagging sharks                                          4
boat swamped                                            4
air disaster                                            4
lifesaving drill                                        4
feeding sharks                                          4
clamming                                                4
floating on a raft                                      3
unknown                                                 3
wreck of the schooner pohoiki                           3
standing in knee-deep water                             3
escaping from alacatraz                                 3
sculling                                                3
competing in the woodvale atlantic rowing race          3
crabbing                                                3
shark watching                                          3
paddling                                                3
sitting on gunwale of boat                              3
suicide                                                 3
sup                                                     3
bathing                                                 3
their 9 m launch was run down by a 25,000-ton j...      3
.                                                       3
rowing                                                  2
sitting in shallows                                     2
shark tagging                                           2
floating in inner tube                                  2
netting shrimp                                          2
aircraft crashed into sea                               2
collecting shells                                       2
shooting sharks                                         2
beach seine netting                                     2
boat with 5 men capsized while returning to the...      2
bathing in river                                        2
floating on back                                        2
hmas parramatta torpedoed & sunk by the u-559           2
standing                                          

In [58]:
actividades_new = pd.DataFrame(dc.activity.value_counts())
actividades_new.shape

(642, 1)

Hemos conseguido bajar el tipo de actividades a 1/3 de la cantidad original.

In [65]:
dc.head(10)

casenumber         date    year        type    country             area  \
0    2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1    2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2    2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3    2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4    2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   
5  2018.06.03.b  03-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
6  2018.06.03.a  03-Jun-2018  2018.0  Unprovoked     BRAZIL       Pernambuco   
7    2018.05.27  27-May-2018  2018.0  Unprovoked        USA          Florida   
8  2018.05.26.b  26-May-2018  2018.0  Unprovoked        USA          Florida   
9  2018.05.26.a  26-May-2018  2018.0  Unprovoked        USA          Florida   

                                            location  activity  \
0                        Oceanside, San Diego County  paddling   
1                     St. Simon Island, Glynn County  standing   
2                                       Habush, Oahu   surfing   
3                                 Arrawarra Headland   surfing   
4                                           La Ticla    diving   
5                                 Flat Rock, Ballina   surfing   
6                              Piedade Beach, Recife  swimming   
7  Lighhouse Point Park, Ponce Inlet, Volusia County   fishing   
8                       Cocoa Beach, Brevard  County   walking   
9                      Daytona Beach, Volusia County  standing   

                     name sex  age  \
0             Julie Wolfe   F   57   
1         Adyson McNeely    F   11   
2             John Denges   M   48   
3                    male   M  NaN   
4          Gustavo Ramos    M  NaN   
5                 Chris 
   M  NaN   
6  Jose Ernesto da Silva    M   18   
7                    male   M   52   
8               Cody High   M   15   
9                    male   M   12   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   
5                            No injury, board bitten          N   
6                                              FATAL          Y   
7            Minor injury to foot. PROVOKED INCIDENT          N   
8                              Lower left leg bitten          N   
9                               Minor injury to foot          N   

             time          species              investigatororsource  \
0           18h00      White shark                  R. Collier, GSAF   
1   14h00  -15h00              NaN    K.McMurray, TrackingSharks.com   
2           07h45              NaN    K.McMurray, TrackingSharks.com   
3             NaN        2 m shark                    B. Myatt, GSAF   
4             NaN  Tiger shark, 3m                         A .Kipper   
5             NaN              NaN         Daily Telegraph, 6/4/2018   
6  Late afternoon      Tiger shark    Diario de Pernambuco, 6/4/2018   
7             NaN  Lemon shark, 3'   K. McMurray, TrackingSharks.com   
8           17h00   Bull shark, 6'    K.McMurray, TrackingSharks.com   
9           14h00              NaN  K. McMurray, Tracking Sharks.com   

                             pdf  \
0           2018.06.25-Wolfe.pdf   
1         2018.06.18-McNeely.pdf   
2          2018.06.09-Denges.pdf   
3       2018.06.08-Arrawarra.pdf   
4           2018.06.04-Ramos.pdf   
5      2018.06.03.b-FlatRock.pdf   
6       2018.06.03.a-daSilva.pdf   
7           2018.05.27-Ponce.pdf   
8          2018.05.26.b-High.pdf   
9  2018.05.26.a-DaytonaBeach.pdf   

                                         hrefformu

#### Limpieza columna `species`
Vamos a cambiar los valores nan de la columna por unkonwn

In [66]:
dc.species[dc.species.isna()] # Empezamos mostrando los valores nulos que hay en la columna

1       NaN
2       NaN
5       NaN
9       NaN
11      NaN
12      NaN
15      NaN
17      NaN
19      NaN
27      NaN
28      NaN
29      NaN
41      NaN
54      NaN
61      NaN
62      NaN
67      NaN
68      NaN
69      NaN
70      NaN
81      NaN
83      NaN
84      NaN
85      NaN
88      NaN
90      NaN
91      NaN
92      NaN
94      NaN
95      NaN
97      NaN
99      NaN
106     NaN
108     NaN
109     NaN
110     NaN
111     NaN
119     NaN
126     NaN
127     NaN
130     NaN
131     NaN
132     NaN
135     NaN
141     NaN
142     NaN
144     NaN
146     NaN
147     NaN
148     NaN
149     NaN
150     NaN
152     NaN
154     NaN
157     NaN
158     NaN
160     NaN
161     NaN
162     NaN
163     NaN
164     NaN
165     NaN
167     NaN
168     NaN
173     NaN
174     NaN
175     NaN
176     NaN
181     NaN
182     NaN
184     NaN
185     NaN
188     NaN
190     NaN
193     NaN
194     NaN
200     NaN
201     NaN
202     NaN
204     NaN
205     NaN
207     NaN
208     NaN
209 

In [68]:
# Podemos comprobar que hay bastante valor nulo, es natural que no todas las víctimas de ataque desconozcan las especies de tiburón
dc.species.fillna('unknown', inplace=True)

In [70]:
dc.head()

casenumber         date    year        type    country             area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         location  activity             name sex  age  \
0     Oceanside, San Diego County  paddling      Julie Wolfe   F   57   
1  St. Simon Island, Glynn County  standing  Adyson McNeely    F   11   
2                    Habush, Oahu   surfing      John Denges   M   48   
3              Arrawarra Headland   surfing             male   M  NaN   
4                        La Ticla    diving   Gustavo Ramos    M  NaN   

                                              injury fatal(y/n)  \
0  No injury to occupant, outrigger canoe and pad...          N   
1                         Minor injury to left thigh          N   
2       Injury to left lower leg from surfboard skeg          N   
3                          Minor injury to lower leg          N   
4  Lacerations to leg & hand shark PROVOKED INCIDENT          N   

            time          species            investigatororsource  \
0          18h00      White shark                R. Collier, GSAF   
1  14h00  -15h00          unknown  K.McMurray, TrackingSharks.com   
2          07h45          unknown  K.McMurray, TrackingSharks.com   
3            NaN        2 m shark                  B. Myatt, GSAF   
4            NaN  Tiger shark, 3m                       A .Kipper   

                        pdf  \
0      2018.06.25-Wolfe.pdf   
1    2018.06.18-McNeely.pdf   
2     2018.06.09-Denges.pdf   
3  2018.06.08-Arrawarra.pdf   
4      2018.06.04-Ramos.pdf   

                                         hrefformula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href casenumber.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   2018.06.04   

  casenumber.2  originalorder unnamed:22 unnamed:23  
0   2018.06.25         6303.0        NaN        NaN  
1   2018.06.18         6302.0        NaN        NaN  
2   2018.06.09         6301.0        NaN        NaN  
3   2018.06.08         6300.0        NaN        NaN  
4   2018.06.04         6299.0        NaN        NaN